lib/Conversion/TTIRToTTMetal/TTIRToTTMetal.cpp

```diff
@@ -107,42 +118,50 @@ class TTIRToTTMetalLayoutRewriter : public OpRewritePattern<ttir::ToLayoutOp> {
   //
   // The return value is a map of physical cores where each core has
   // an associated list of noc reads/writes to be performed.
-  llvm::MapVector<PhysicalCoreCoord, mlir::SmallVector<NocTx>>
-  calculateDataMovement(ArrayRef<int64_t> tensorShape, std::int64_t elemSize,
-                        AffineMap src, AffineMap dst, NocTx::Type type) const {
+  static llvm::MapVector<PhysicalCoreCoord, mlir::SmallVector<NocTx>>
+  calculateDataMovement(ArrayRef<int64_t> tensorShape, AffineMap srcToDstShape,
+                        std::int64_t elemSize, AffineMap src, AffineMap dst,
+                        NocTx::Type type, std::int64_t dstCapacity) {
     bool read = type == NocTx::Type::Read;
     llvm::MapVector<PhysicalCoreCoord, mlir::SmallVector<NocTx>> txMap;
     assert(src.getNumResults() == MemoryMapResultIdx::NumIndices);
     assert(dst.getNumResults() == MemoryMapResultIdx::NumIndices);
 
-    ::ttmlir::utils::sample(tensorShape, [&txMap, src, dst, elemSize, read,
-                                          type](ArrayRef<std::int64_t> index) {
-      SmallVector<int64_t> srcResults = src.compose(index);
-      SmallVector<int64_t> dstResults = dst.compose(index);
-      assert(srcResults.size() == src.getNumResults());
-      assert(dstResults.size() == dst.getNumResults());
-      PhysicalCoreCoord srcCoord(srcResults);
-      PhysicalCoreCoord dstCoord(dstResults);
-      std::int64_t srcOffset = srcResults.back() * elemSize;
-      std::int64_t dstOffset = dstResults.back() * elemSize;
-      SmallVector<NocTx> &txs = txMap[read ? dstCoord : srcCoord];
-      if (not txs.empty() && txs.back().isContiguous(read ? srcCoord : dstCoord,
-                                                     srcOffset, dstOffset)) {
-        txs.back().size += elemSize;
-      } else {
-        txs.push_back(NocTx(type, read ? srcCoord : dstCoord, srcOffset,
-                            dstOffset, elemSize));
-      }
-    });
+    ::ttmlir::utils::sample(
+        tensorShape, [&txMap, &srcToDstShape, src, dst, elemSize, read, type,
+                      dstCapacity](ArrayRef<std::int64_t> index) {
+          SmallVector<int64_t> srcResults = src.compose(index);
+          SmallVector<int64_t> dstResults =
+              dst.compose(srcToDstShape.compose(index));
+          assert(srcResults.size() == src.getNumResults());
+          assert(dstResults.size() == dst.getNumResults());
+          PhysicalCoreCoord srcCoord(srcResults);
+          PhysicalCoreCoord dstCoord(dstResults);
+          std::int64_t srcOffset = srcResults.back() * elemSize;
+          std::int64_t dstOffset = dstResults.back() * elemSize;
+
+          SmallVector<NocTx> &txs = txMap[read ? dstCoord : srcCoord];
+          if (not txs.empty() &&
+              txs.back().isContiguous(read ? srcCoord : dstCoord, srcOffset,
+                                      dstOffset) &&
+              txs.back().size + elemSize <= dstCapacity) {
```

Comment:
Yeah I agree. I haven't enforced all the needed rules to avoid such deadlocks. I was going to think about it when we start sizing CBs differently from initial tensor as part of https://github.com/tenstorrent/tt-mlir/issues/778.

We must respect this rule for reserve_back/wait_front calls.
>CB total size must be an even multiple of the argument passed to this call.

Therefore I think your idea of breaking down TXs is good. 

I am just little confused by wording "even multiple". Does this mean CB size must be 
1. even number & multiple of the argument? or
2. (CB_size / arg) % 2 == 0. 

I believe it is the first one, as it makes more sense, because the second one would forbid reserving 4 tiles when CB size is 12.

Reply:
Yes sorry, I meant just _multiple of_.  By even I meant _evenly divisible_.

Reply:
> I was going to think about it when we start sizing CBs differently from initial tensor as part of https://github.com/tenstorrent/tt-mlir/issues/778.

I need to think a bit more about this, naively it feels like a pretty complicated function to solve because you'd need to model how the reads will actually be coalesced, just as we do in this function.  You could imagine my example is of 3 cores with blocks of 2x2 tiles, if the read order happens to be row major then we read all 4 tiles in one go and we're in trouble, if the read order was col major then we're actually OK because we ended up just gathering 1 tile at a time.

Reply:
1. You did not confuse me, I was confused by docstring from `cb_api.h` where `even multiple` term is mentioned. I'll just consider this to be even number which is divisible by the arg. So 12 & 4 are valid numbers.

2. 
> You could imagine my example above is of 3 cores with blocks of 2x2 tiles, if the read order happens to be row major then we read all 4 tiles in one go and we're in trouble,

Not sure I got this part. 
My assumptions are that if kernel consumes 6 units - meaning it does `cb_wait_front(in_cb, 6)`, we can't coalesce read TX to more than 3 units. 

----------

Anyway, if you want we can continue discussion on this topic in #778, so I can wrap this PR up. What do you say?

Reply:
Yeah let's follow up on #778